rotkehlchen/chain/evm/decoding/odos/v2/decoder.py

```diff
@@ -0,0 +1,228 @@
+                event.event_type = HistoryEventType.TRADE
+                event.address = self.router_address
+                event.counterparty = CPT_ODOS_V2
+                out_events.append(event)
+
+            elif output_tokens.get(event.asset.identifier) == event.balance.amount and (
+                (
+                    event.event_type == HistoryEventType.RECEIVE and
+                    event.event_subtype == HistoryEventSubType.NONE
+                ) or (
+                    event.event_type == HistoryEventType.TRADE and
+                    event.event_subtype == HistoryEventSubType.RECEIVE
+                )
+            ):
+                event.event_type = HistoryEventType.TRADE
+                event.event_subtype = HistoryEventSubType.RECEIVE
+                event.notes = f'Receive {event.balance.amount} {event.asset.symbol_or_name()} as the result of a swap in Odos v1'  # noqa: E501
+                event.address = self.router_address
+                event.counterparty = CPT_ODOS_V2
+                in_events.append(event)
+
+        # fees are the amount of out tokens that router received - the amount that router sent
+        router_holding_amount: defaultdict[Asset, FVal] = defaultdict(FVal)
+        for tx_log in context.all_logs:
+            if (
+                tx_log.topics[0] != ERC20_OR_ERC721_TRANSFER or
+                ((identifier := evm_address_to_identifier(
+                    address=tx_log.address,
+                    chain_id=self.evm_inquirer.chain_id,
+                    token_type=EvmTokenKind.ERC20,
+                )) not in output_tokens)
+            ):
+                continue
+
+            # can directly initialize EvmToken here because it's in output_tokens which were get_or_created above  # noqa: E501
+            asset = EvmToken(identifier)
+
+            if hex_or_bytes_to_address(tx_log.topics[2]) == self.router_address:
+                router_holding_amount[asset] += FVal(hex_or_bytes_to_int(tx_log.data)) / FVal(10 ** asset.get_decimals())  # noqa: E501
+
+            elif hex_or_bytes_to_address(tx_log.topics[1]) == self.router_address:
+                router_holding_amount[asset] -= FVal(hex_or_bytes_to_int(tx_log.data)) / FVal(10 ** asset.get_decimals())  # noqa: E501
+
+        if self.native_currency.identifier in output_tokens:
+            try:
+                internal_native_in = self.evm_txns.get_and_ensure_internal_txns_of_parent_in_db(
+                    tx_hash=context.transaction.tx_hash,
+                    to_address=self.router_address,
+                    user_address=decoded_data[0],
+                )
+                internal_native_out = self.evm_txns.get_and_ensure_internal_txns_of_parent_in_db(
+                    tx_hash=context.transaction.tx_hash,
+                    from_address=self.router_address,
+                    to_address=decoded_data[0],
+                    user_address=decoded_data[0],
+                )
```

Comment:
these are to check if any native currency was moved in/out of the router, they are treated differently because they are not ERC20 transfers and only happen through internal transactions